README
NAME
    Math::Random::MT - The Mersenne Twister PRNG

SYNOPSIS
      ## Object-oriented interface:
      use Math::Random::MT;
      $gen = Math::Random::MT->new()        # or...
      $gen = Math::Random::MT->new($seed);  # or...
      $gen = Math::Random::MT->new(@seeds);
      $seed = $gen->get_seed();             # seed used to generate the random numbers
      $rand = $gen->rand(42);               # random number in the interval [0, 42)
      $dice = int($gen->rand(6)+1);         # random integer between 1 and 6
      $coin = $gen->rand() < 0.5 ?          # flip a coin
        "heads" : "tails"

      ## Function-oriented interface:
      use Math::Random::MT qw(srand rand);
      # now use srand() and rand() as you usually do in Perl

DESCRIPTION
    The Mersenne Twister is a pseudorandom number generator developed by
    Makoto Matsumoto and Takuji Nishimura. It is described in their paper at
    <URL:http://www.math.keio.ac.jp/~nisimura/random/doc/mt.ps>. This
    algorithm has a very uniform distribution and is good for modelling
    purposes but do not use it for cryptography.

    This module implements two interfaces:

  Object-oriented interface
    new()
        Creates a new generator that is automatically seeded based on
        gettimeofday.

    new($seed)
        Creates a new generator seeded with an unsigned 32-bit integer.

    new(@seeds)
        Creates a new generator seeded with an array of (up to 624) unsigned
        32-bit integers.

    set_seed()
        Seeds the generator. It takes the same arguments as *new()*.

    get_seed()
        Retrieves the value of the seed used.

    rand($num)
        Behaves exactly like Perl's builtin rand(), returning a number
        uniformly distributed in [0, $num) ($num defaults to 1).

  Function-oriented interface
    rand($num)
        Behaves exactly like Perl's builtin rand(), returning a number
        uniformly distributed in [0, $num) ($num defaults to 1).

    srand($seed)
        Behaves just like Perl's builtin srand(). As in Perl >= 5.14, the
        seed is returned. If you use this interface, it is strongly
        recommended that you call *srand()* explicitly, rather than relying
        on *rand()* to call it the first time it is used.

SEE ALSO
    <URL:http://www.math.sci.hiroshima-u.ac.jp/~m-mat/MT/emt.html>

    Data::Entropy

ACKNOWLEDGEMENTS
    Sean M. Burke
        For giving me the idea to write this module.

    Philip Newton
        For several useful patches.

    Florent Angly
        For implementing seed generation and retrieval.

AUTHOR
    Abhijit Menon-Sen <ams@toroid.org>

    Copyright 2001 Abhijit Menon-Sen. All rights reserved.

    Based on the C implementation of MT19937 Copyright (C) 1997 - 2002,
    Makoto Matsumoto and Takuji Nishimura

    This software is distributed under a (three-clause) BSD-style license.
    See the LICENSE file in the distribution for details.



Changes
1.16    2012-08-06      Florent Angly <florent.angly@gmail.com>

    * Fixed issue introduced in version 1.15 where rand() took no notice of
      argument and irand() did (bug #78200, reported by David Morel)

    * Migration of test suite to Test::More and Test::Number::Delta


1.15    2012-06-04      Florent Angly <florent.angly@gmail.com>

    * Implemented irand() to draw random integers (bug #73298, feature requested
      by crew@cs.stanford.edu)

    * Fixed build failure on Hurd (bug #74165, reported and patched by Salvatore
      Bonaccorso)


1.14    2012-05-27      Florent Angly <florent.angly@gmail.com>

    * Fixed compilation problems with nmake on Windows platforms
      (Florent Angly, bug #74984)

    * Fixed issues with random seed (bug #77343, solution suggested by
      Laurent Dami, fix implemented by Florent Angly)


1.13    2012-01-20      Abhijit Menon-Sen <ams@toroid.org>

    * New seed generation and retrieval mechanism by Florent Angly (bug
      #64640 on rt.cpan.org).


1.12    2010-09-29      Abhijit Menon-Sen <ams@toroid.org>

    * Fix OS X build failure reported by Daniel Barker and fixed by
      Tokuhiro Matsuno.


1.11    2010-05-08      Abhijit Menon-Sen <ams@toroid.org>

    * Fix tests broken by overly optimistic floating point comparison
      and long doubles in new releases of Perl. (Thanks to Salvatore
      Bonaccorso for the report and fix.)


1.10    2009-08-06      Abhijit Menon-Sen <ams@toroid.org>

    * Compile fix for MSVC, which has no stdint.h.
      (Thanks to Taro Nishino for the report and patch.)

    (I made a mistake releasing 1.09.)


1.08    2009-08-06      Abhijit Menon-Sen <ams@toroid.org>

    * BSD license

    * Compile fix for GNU/kFreeBSD from Cyril Brulebois (via Debian)


1.07    2007-01-12      Abhijit Menon-Sen <ams@toroid.org>

    * Add the Changes file to the MANIFEST, in addition to editing it
      each time. (Thanks to John M. Gamble.)

      There have been no changes in functionality since 1.03.


1.06    2007-01-11      Abhijit Menon-Sen <ams@toroid.org>

    * Minor POD fix from Steven Schubiger.


1.05    2007-01-11      Abhijit Menon-Sen <ams@toroid.org>

    * Mingw compile fix from Steffen Mueller.


1.04    2004-10-09      Abhijit Menon-Sen <ams@wiw.org>

    * Incorporated some compile fixes from Gordon Lack.

    (1.03 had only some documentation changes.)


1.02    2004-05-07      Abhijit Menon-Sen <ams@wiw.org>

    * Added support for seeding the random number generator with an
      array of numbers, as suggested by Philip M. Feldman.

    (1.01 was not released.)


1.00    2002-02-09      Abhijit Menon-Sen <ams@wiw.org>

    * First stable release.


_mt.c
#include "mt.h"

#include <stdlib.h>
#include <stdio.h>

/* This code is based on mt19937ar.c, written by Takuji Nishimura and
   Makoto Matsumoto (20020126). Further details are available at
   <http://www.math.sci.hiroshima-u.ac.jp/~m-mat/MT/emt.html>.

   REFERENCE
   M. Matsumoto and T. Nishimura,
   "Mersenne Twister: A 623-Dimensionally Equidistributed Uniform
   Pseudo-Random Number Generator",
   ACM Transactions on Modeling and Computer Simulation,
   Vol. 8, No. 1, January 1998, pp 3--30. */

struct mt *mt_init(void)
{
    struct mt *self = malloc(sizeof(struct mt));
    return self;
}

void mt_free(struct mt *self)
{
    free(self);
}

uint32_t mt_get_seed(struct mt *self)
{
    return self->seed;
}

void mt_init_seed(struct mt *self, uint32_t seed)
{
    int i;
    uint32_t *mt;

    mt = self->mt;
    mt[0] = seed & 0xffffffff;

    for ( i = 1; i < N; i++ )
        mt[i] = 1812433253 * (mt[i-1]^(mt[i-1]>>30)) + i;
    self->mti = N;
    self->seed = mt[0];
}

void mt_setup_array(struct mt *self, uint32_t *array, int n)
{
    int i, j, k;
    uint32_t *mt;

    mt_init_seed( self, 19650218UL );

    i = 1; j = 0;
    k = ( N > n ? N : n );
    mt = self->mt;

    for (; k; k--) {
        mt[i] = (mt[i] ^ ((mt[i-1] ^ (mt[i-1] >> 30)) * 1664525UL)) + array[j]
                + j;
        i++; j++;
        if (i>=N) { mt[0] = mt[N-1]; i=1; }
        if (j>=n) j=0;
    }
    for (k=N-1; k; k--) {
        mt[i] = (mt[i] ^ ((mt[i-1] ^ (mt[i-1] >> 30)) * 1566083941UL)) - i;
        i++;
        if (i>=N) { mt[0] = mt[N-1]; i=1; }
    }

    mt[0] = 0x80000000UL;
}

/* Returns a pseudorandom number which is uniformly distributed in [0,1) */

double mt_genrand(struct mt *self)
{
    int kk;
    uint32_t y;
    static uint32_t mag01[2] = {0x0, 0x9908b0df};
    static const uint32_t UP_MASK = 0x80000000, LOW_MASK = 0x7fffffff;

    if (self->mti >= N) {
        for (kk = 0; kk < N-M; kk++) {
            y = (self->mt[kk] & UP_MASK) | (self->mt[kk+1] & LOW_MASK);
            self->mt[kk] = self->mt[kk+M] ^ (y >> 1) ^ mag01[y & 1];
        }

        for (; kk < N-1; kk++) {
            y = (self->mt[kk] & UP_MASK) | (self->mt[kk+1] & LOW_MASK);
            self->mt[kk] = self->mt[kk+(M-N)] ^ (y >> 1) ^ mag01[y & 1];
        }

        y = (self->mt[N-1] & UP_MASK) | (self->mt[0] & LOW_MASK);
        self->mt[N-1] = self->mt[M-1] ^ (y >> 1) ^ mag01[y & 1];

        self->mti = 0;
    }
  
    y  = self->mt[self->mti++];
    y ^= y >> 11;
    y ^= y <<  7 & 0x9d2c5680;
    y ^= y << 15 & 0xefc60000;
    y ^= y >> 18;

    return y*(1.0/4294967296.0);
}


MT.pm
package Math::Random::MT;

use strict;
use Carp;
use DynaLoader;
use vars qw( @ISA $VERSION );

my $gen = undef;
@ISA = qw( DynaLoader );
$VERSION = '1.16';

bootstrap Math::Random::MT $VERSION;

sub new
{
    my ($class, @seeds) = @_;

    my $self = Math::Random::MT::init();
    $self->set_seed(@seeds);

    return $self;
}

sub set_seed
{
    my ($self, @seeds) = @_;
    @seeds > 1 ? $self->setup_array(@seeds) :
                 $self->init_seed($seeds[0] || _rand_seed());
    return $self->get_seed;
}

sub srand
{
    my (@seeds) = @_;
    $gen = Math::Random::MT->new(@seeds);
    return $gen->get_seed;
}

sub rand
{
    my ($self, $N) = @_;

    unless (ref $self) {
        $N = $self;
        Math::Random::MT::srand() unless defined $gen;
        $self = $gen;
    }

    return ($N || 1) * $self->genrand();
}

sub irand
{
    my ($self) = @_;

    unless (ref $self) {
        Math::Random::MT::srand() unless defined $gen;
        $self = $gen;
    }

    return $self->genirand();
}


# Generate a random seed using the built-in PRNG.

sub _rand_seed {
    my ($self) = @_;

    # Get a seed at random through Perl's CORE::rand(). We do not call
    # CORE::srand() to avoid altering the random numbers that other parts of
    # the running script might be using. The seeds obtained by rapid calls to
    # the _rand_seed() function are all different.
    
    return int(CORE::rand(2**32));
}

sub import
{
    no strict 'refs';
    my $pkg = caller;
    foreach my $sym (@_) {
        if ($sym eq 'srand' || $sym eq 'rand' || $sym eq 'irand') {
            *{"${pkg}::$sym"} = \&$sym;
        }
    }
}

1;

__END__

=head1 NAME

Math::Random::MT - The Mersenne Twister PRNG

=head1 SYNOPSIS

  ## Object-oriented interface:
  use Math::Random::MT;
  $gen = Math::Random::MT->new()        # or...
  $gen = Math::Random::MT->new($seed);  # or...
  $gen = Math::Random::MT->new(@seeds);
  $seed = $gen->get_seed();             # seed used to generate the random numbers
  $rand = $gen->rand(42);               # random number in the interval [0, 42)
  $dice = int($gen->rand(6)+1);         # random integer between 1 and 6
  $coin = $gen->rand() < 0.5 ?          # flip a coin
    "heads" : "tails"
  $int = $gen->irand();                 # random integer in [0, 2^32-1]

  ## Function-oriented interface:
  use Math::Random::MT qw(srand rand irand);
  # now use srand() and rand() as you usually do in Perl

=head1 DESCRIPTION

The Mersenne Twister is a pseudorandom number generator developed by
Makoto Matsumoto and Takuji Nishimura. It is described in their paper at
<URL:http://www.math.keio.ac.jp/~nisimura/random/doc/mt.ps>. This algorithm
has a very uniform distribution and is good for modelling purposes but do not
use it for cryptography.

This module implements two interfaces:

=head2 Object-oriented interface

=over

=item new()

Creates a new generator that is automatically seeded based on gettimeofday.

=item new($seed)

Creates a new generator seeded with an unsigned 32-bit integer.

=item new(@seeds)

Creates a new generator seeded with an array of (up to 624) unsigned
32-bit integers.

=item set_seed()

Seeds the generator and returns the seed used. It takes the same arguments as
I<new()>.

=item get_seed()

Retrieves the value of the seed used.

=item rand($num)

Behaves exactly like Perl's builtin rand(), returning a number uniformly
distributed in [0, $num) ($num defaults to 1).

=item irand()

Returns a 32-bit integer, i.e. an integer uniformly distributed in [0, 2^32-1].

=back

=head2 Function-oriented interface

=over

=item srand($seed)

Behaves just like Perl's builtin srand(). As in Perl >= 5.14, the seed is
returned. If you use this interface, it is strongly recommended that you
call I<srand()> explicitly, rather than relying on I<rand()> to call it the
first time it is used.

=item rand($num)

Behaves exactly like Perl's builtin rand(), returning a number uniformly
distributed in [0, $num) ($num defaults to 1).

=item irand()

Returns a 32-bit integer, i.e. an integer uniformly distributed in [0, 2^32-1].

=back

=head1 SEE ALSO

<URL:http://www.math.sci.hiroshima-u.ac.jp/~m-mat/MT/emt.html>

Data::Entropy

=head1 ACKNOWLEDGEMENTS

=over 4

=item Sean M. Burke

For giving me the idea to write this module.

=item Philip Newton

For several useful patches.

=item Florent Angly

For implementing seed generation and retrieval.

=back

=head1 AUTHOR

Abhijit Menon-Sen <ams@toroid.org>

Copyright 2001 Abhijit Menon-Sen. All rights reserved.

Based on the C implementation of MT19937
Copyright (C) 1997 - 2002, Makoto Matsumoto and Takuji Nishimura

This software is distributed under a (three-clause) BSD-style license.
See the LICENSE file in the distribution for details.
